src/Stratis.Bitcoin.Features.PoA/Voting/IdleFederationMembersKicker.cs

```diff
@@ -0,0 +1,144 @@
+
+            if (this.fedMembersByLastActiveTime == null)
+            {
+                this.logger.LogDebug("No saved data found. Initializing federation data with genesis timestamps.");
+
+                this.fedMembersByLastActiveTime = new Dictionary<PubKey, uint>();
+
+                // Initialize federation members with genesis time.
+                foreach (PubKey federationMember in this.federationManager.GetFederationMembers())
+                    this.fedMembersByLastActiveTime.Add(federationMember, this.network.GenesisTime);
+
+                this.SaveMembersByLastActiveTime();
+            }
+        }
+
+        private void OnFedMemberKicked(FedMemberKicked fedMemberKickedData)
+        {
+            this.fedMembersByLastActiveTime.Remove(fedMemberKickedData.KickedMember);
+
+            this.SaveMembersByLastActiveTime();
+        }
+
+        private void OnFedMemberAdded(FedMemberAdded fedMemberAddedData)
+        {
+            if (!this.fedMembersByLastActiveTime.ContainsKey(fedMemberAddedData.AddedMember))
+            {
+                this.fedMembersByLastActiveTime.Add(fedMemberAddedData.AddedMember, this.consensusManager.Tip.Header.Time);
+
+                this.SaveMembersByLastActiveTime();
+            }
+        }
+
+        private void OnBlockConnected(BlockConnected blockConnectedData)
+        {
+            // Update last active time.
+            uint timestamp = blockConnectedData.ConnectedBlock.ChainedHeader.Header.Time;
+            PubKey key = this.slotsManager.GetPubKeyForTimestamp(timestamp);
+            this.fedMembersByLastActiveTime.AddOrReplace(key, timestamp);
+
+            this.SaveMembersByLastActiveTime();
+
+            // Check if any fed member was idle for too long.
+            ChainedHeader tip = this.consensusManager.Tip;
+
+            foreach (KeyValuePair<PubKey, uint> fedMemberToActiveTime in this.fedMembersByLastActiveTime)
+            {
+                uint inactiveForSeconds = tip.Header.Time - fedMemberToActiveTime.Value;
+
+                if (inactiveForSeconds > this.federationMemberMaxIdleTimeSeconds)
+                {
+                    this.logger.LogDebug("Federation member '{0}' was inactive for {1} seconds and will be scheduled to be kicked.", fedMemberToActiveTime.Key, inactiveForSeconds);
+
+                    this.votingManager.ScheduleVote(new VotingData()
+                    {
+                        Key = VoteKey.KickFederationMember,
+                        Data = fedMemberToActiveTime.Key.ToBytes()
+                    });
+                }
+            }
+        }
+
+        private void SaveMembersByLastActiveTime()
+        {
+            this.keyValueRepository.SaveValueJson(fedMembersByLastActiveTimeKey, this.fedMembersByLastActiveTime);
+        }
+
+        /// <inheritdoc />
+        public void Dispose()
+        {
+            if (this.blockConnectedToken != null)
```

Comment:
During the eventbus PR i noticed that some tests were failing because the code tried to unsubscribe events that wasn't registered (token null). Probably it happened because dispose was called twice or because was called before Initialize ended (not sure if it's possible), anyway I did a change to just ignore Unsubscribe calls with a null argument, so you can save this check

Reply:
I think there should be a null check and event bus should throw if null is provided. 
Also it should be possible to execute Dispose twice with no negative consequence (I think it's covered in dispose docs from microsoft)

Reply:
I don't see why throwing for null unsubscribe, it has not side effects because it means you didn't registered in first place.

The only useful thing I find about this is that it may show you that you forget to store a subscription token in the variable you are passing, and we can add it if needed, but first then we must find the reason why integration tests fails (it doesn't happen on integration tests run locally)

And if you want to have proper checks, then it's not enough to just check for blockConnectedToken != null becuse you should do the check for every subscription token

Reply:
about MS advices, If you stick with its implementation then you need to implement IDispose like this

```
#region IDisposable Support
private bool disposedValue = false; // To detect redundant calls

protected virtual void Dispose(bool disposing) {
   if (!disposedValue) {
      if (disposing) {
         // TODO: dispose managed state (managed objects).
      }

      // TODO: free unmanaged resources (unmanaged objects) and override a finalizer below.
      // TODO: set large fields to null.

      disposedValue = true;
   }
}

// TODO: override a finalizer only if Dispose(bool disposing) above has code to free unmanaged resources.
// ~X() {
//   // Do not change this code. Put cleanup code in Dispose(bool disposing) above.
//   Dispose(false);
// }

// This code added to correctly implement the disposable pattern.
public void Dispose() {
   // Do not change this code. Put cleanup code in Dispose(bool disposing) above.
   Dispose(true);
   // TODO: uncomment the following line if the finalizer is overridden above.
   // GC.SuppressFinalize(this);
}
#endregion
```

Reply:
>And if you want to have proper checks, then it's not enough to just check for blockConnectedToken != null becuse you should do the check for every subscription token

blockConnectedToken and other tokens are set at the initialization. So if blockConnectedToken is not null- rest of them are not null too. So it's sufficient to just check 1.

Reply:
>about MS advices, If you stick with its implementation then you need to implement IDispose like this

I do not plan to override dispose. So I don't see why we need that overload.

Reply:
that's the base implementation offered by visual studio snipped, virtual can be omitted, I was just replying to your 
> Also it should be possible to execute Dispose twice with no negative consequence (I think it's covered in dispose docs from microsoft)

and for be thread-safe this may even need a lock. I'm talking from a generic pov not strictly related to this commit